src/default-implementations.lisp

```diff
@@ -318,6 +318,111 @@ cannot be decremented in that time, returns NIL without decrementing the count."
   "Returns T if OBJECT is a semaphore; returns NIL otherwise."
   (typep object 'semaphore))
 
+;;; Mailboxes
+
+(defdfun make-mailbox (&key name initial-contents)
+  "Returns a new MAILBOX with messages in INITIAL-CONTENTS enqueued."
+  (check-type name (or null string))
+  (check-type initial-contents list)
+  (let ((locative (list nil)))
+    (make-%mailbox :name name
+                   :lock (make-lock name)
+                   :cvar (make-condition-variable :name name)
+                   :count (length initial-contents)
+                   :messages (append initial-contents locative)
+                   :messages-tail locative)))
+
+(defdfun mailbox-name (mailbox)
+  "Name of a MAILBOX. SETFable."
+  (%mailbox-name mailbox))
+
+(defdfun (setf mailbox-name) (name mailbox)
+  "Sets a MAILBOX name."
+  (setf (%mailbox-name mailbox) name))
+
+(defdfun list-mailbox-messages (mailbox)
+  "Returns a fresh list containing all the messages in the
+mailbox. Does not remove messages from the mailbox."
+  (with-lock-held ((%mailbox-lock mailbox))
+    (loop
+      for (i . j) on (%mailbox-messages mailbox)
+      until (null j) collect i)))
+
+(defdfun mailbox-count (mailbox)
+  "Returns the number of messages currently in the mailbox."
+  (with-lock-held ((%mailbox-lock mailbox))
+    (%mailbox-count mailbox)))
+
+(defdfun mailbox-empty-p (mailbox)
+  "Returns true if MAILBOX is currently empty, NIL otherwise."
+  (with-lock-held ((%mailbox-lock mailbox))
+    (zerop (%mailbox-count mailbox))))
+
+(defdfun receive-message (mailbox &key timeout)
+  "Removes the oldest message from MAILBOX and returns it as the primary
+value, and a secondary value of T. If MAILBOX is empty waits until a message
+arrives.
+
+If TIMEOUT is provided, and no message arrives within the specified interval,
+returns primary and secondary value of NIL."
+  (with-lock-held ((%mailbox-lock mailbox))
```

Comment:
You need to care for the timeout there, don't you?

Reply:
good catch! looking into it.

Reply:
I've fixed both with a rather ugly solution. Semaphore fixup has a new commit while mailbox fixup is squashed to the main commit.

Reply:
I'll put it here for the reference

```
(defdfun wait-on-semaphore (semaphore &key timeout)
  "Decrement the count of SEMAPHORE by 1 if the count would not be negative.

Else blocks until the semaphore can be decremented. Returns generalized boolean
T on success.

If TIMEOUT is given, it is the maximum number of seconds to wait. If the count
cannot be decremented in that time, returns NIL without decrementing the count."
  (let ((deadline (and timeout
                       (+ (get-internal-real-time)
                          (* timeout internal-time-units-per-second))))
        (lock (%semaphore-lock semaphore)))
    (macrolet ((with-thunk (() &body body)
                 `(if timeout
                      (handler-case (with-timeout (timeout)
                                      (acquire-lock lock))
                        (timeout ()
                          nil)
                        (:no-error ()
                          (unwind-protect (progn ,@body)
                            (release-lock lock))))
                      (with-lock-held (lock)
                        ,@body))))
      (with-thunk ()
        (if (>= (%semaphore-counter semaphore) 1)
            (decf (%semaphore-counter semaphore))
            (progn
              ;; we need this loop because of a spurious wakeup possibility
              (loop until (>= (%semaphore-counter semaphore) 1)
                    do (cond
                         ((null (condition-wait (%semaphore-condition-variable semaphore)
                                                (%semaphore-lock semaphore)
                                                :timeout timeout))
                          (return-from wait-on-semaphore))
                         ;; unfortunately cv-wait may return T on timeout too
                         ((and deadline (>= (get-internal-real-time) deadline))
                          (return-from wait-on-semaphore))
                         (timeout
                          (setf timeout (/ (- deadline (get-internal-real-time))
                                           internal-time-units-per-second)))))
              (decf (%semaphore-counter semaphore))))))))
```

After a thought I don't think we should put a timeout above the lock. We do not block inside locks so they should finish very quickly (condition-wait releases the lock). This time may be neglected without a consequence I think.

@sionescu  what do you think?